lib/LibraryTemplatePlugin.js

```diff
@@ -125,29 +129,48 @@ class LibraryTemplatePlugin {
 					break;
 				case "amd": {
 					const AmdMainTemplatePlugin = require("./AmdMainTemplatePlugin");
-					new AmdMainTemplatePlugin(this.name).apply(compilation);
+					new AmdMainTemplatePlugin(this._withName(name => name)).apply(
+						compilation
+					);
 					break;
 				}
 				case "umd":
 				case "umd2": {
 					const UmdMainTemplatePlugin = require("./UmdMainTemplatePlugin");
-					new UmdMainTemplatePlugin(this.name, {
-						optionalAmdExternalAsGlobal: this.target === "umd2",
+					new UmdMainTemplatePlugin(this._withName(name => name), {
+						optionalAmdExternalAsGlobal:
+							this._withTarget(target => target) === "umd2",
 						namedDefine: this.umdNamedDefine,
 						auxiliaryComment: this.auxiliaryComment
 					}).apply(compilation);
 					break;
 				}
 				case "jsonp": {
 					const JsonpExportMainTemplatePlugin = require("./web/JsonpExportMainTemplatePlugin");
-					new JsonpExportMainTemplatePlugin(this.name).apply(compilation);
+					new JsonpExportMainTemplatePlugin(this._withName(name => name)).apply(
+						compilation
+					);
 					break;
 				}
 				default:
-					throw new Error(`${this.target} is not a valid Library target`);
+					throw new Error(
+						`${this._withTarget(
+							target => target
+						)} is not a valid Library target`
+					);
 			}
 		});
 	}
+
+	_withName(fn) {
+		return typeof this.name === "function" ? fn(this.name()) : fn(this.name);
+	}
+
+	_withTarget(fn) {
```

Comment:
Please revert all changes regarding `target`. The current approach doesn't work for `target` as function, since we can't apply different plugins per chunk.

Reply:
Concentrate on allow functions for `name`.